### PR TITLE
Make reference panel play nicely with browser history

### LINKED ISSUE #40969: CodeMirror blob: make popover more sticky
Currently, the popover quickly disappears when the mouse hovers around the borders. We should increase the border radius of the hover region to make it easier to move the mouse to definition/references buttons 

Original report https://sourcegraph.slack.com/archives/C03CSAER9LK/p1661556219158469

### PR DESCRIPTION
**Pull request by olafurpg**
Fixes #40969

Loom recording explaining the motivation for this change https://www.loom.com/share/cbbeb6208e5a4046abfec74a55beafbf

Previously, it was easy to loose track of your context while drilling
into definitions/references through the reference panel. The root cause
of the problem is that the reference panel had its own history but there
were no UI components to navigate through that history.

Now, the reference panel no longer supports deep history. Every browser
URL is mapped with a single ref-panel URL (the focused line in the
preview pane) allowing the ref-panel to update accordingly when the
user clicks on the browser's back/forward buttons.

## Test plan

See the CI go green.

## App preview:

- [Web](https://sg-web-olafurpg-ref-preview-codeintel.onrender.com)
- [Storybook](https://5f0f381c0e50750022dc6bf7-vwdlcanewt.chromatic.com)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

**Review by mrnugget — COMMENTED** [review]
Just played around with this a bit. The marking of active location doesn't seem to work consistently:

https://user-images.githubusercontent.com/1185253/191051912-671a919b-bd74-41de-ad43-6023156c3e54.mp4

**Comment by mrnugget** [comment]
Quoting myself:

> The marking of active location doesn't seem to work consistently:

I think I just figured out when it doesn't work: if the clicked location is in anothe repo?

File: client/web/src/repo/blob/Blob.tsx [hunk 1]
```diff
@@ -790,6 +807,29 @@ export const Blob: React.FunctionComponent<React.PropsWithChildren<BlobProps>> =
         }
     }, [codeViewElements])
 
+    // Add the `.clickable-row` CSS class to all rows to give visual hints that they're clickab.e
```
[the remark this answers is not in the document]
Comment (olafurpg) [reply client/web/src/repo/blob/Blob.tsx]: Done

File: client/web/src/repo/blob/Blob.tsx [hunk 2]
```diff
@@ -440,7 +443,14 @@ export const Blob: React.FunctionComponent<React.PropsWithChildren<BlobProps>> =
                         const parameters = new URLSearchParams(location.search)
                         parameters.delete('popover')
 
-                        if (position && !('character' in position)) {
+                        const notBlankSpace = position && !('character' in position)
```
[the remark this answers is not in the document]
Comment (olafurpg) [reply client/web/src/repo/blob/Blob.tsx]: Good point. I agree the logic is convoluted, I refactored the branching logic to be easier to follow.

File: client/web/src/repo/blob/CodeMirrorBlob.tsx [hunk 3]
```diff
@@ -103,13 +108,15 @@ export const Blob: React.FunctionComponent<BlobProps> = props => {
...
         onHandleFuzzyFinder,
+        navigateToLineOnAnyClick,
     } = props
 
     const [container, setContainer] = useState<HTMLDivElement | null>(null)
     // This is used to avoid reinitializing the editor when new locations in the
     // same file are opened inside the reference panel.
     const blobInfo = useDistinctBlob(props.blobInfo)
     const position = useMemo(() => parseQueryAndHash(location.search, location.hash), [location.search, location.hash])
+    // const position = parseQueryAndHash(location.search, location.hash)
```
[the remark this answers is not in the document]
Comment (olafurpg) [reply client/web/src/repo/blob/CodeMirrorBlob.tsx:119]: Removed this line along with other commented out code.

**Review by olafurpg — COMMENTED** [review]
@mrnugget thank you for the review! I was able to reproduce and fix the bug where selected items didn't get highlight. This happened when the reference item in the tree view had the default revision `''` while the preview pane had the commit ID. 

I also cleaned up the code a bit, removed dead comments and unused code including `panelHistory`.

I also added a minor improvement, the highlighted item is now scrolled into view on the first load (updating browser history, or reloading the page). Before this improvement, the preview pane (right side) could point to a specific line of code while the left side would be scrolled at the top.

File: client/web/src/codeintel/ReferencesPanel.tsx [hunk 4]
```diff
@@ -602,52 +612,74 @@ const CollapsibleLocationList: React.FunctionComponent<
...
+    //                         panelHistory.push(location)
+    //                         break
+    //                     case 'REPLACE':
+    //                         panelHistory.replace(location)
+    //                         break
+    //                 }
+    //             }
+    //         }),
+    //     [useCodeMirror, history, panelHistory]
+    // )
```
[the remark this answers is not in the document]
Comment (olafurpg) [reply client/web/src/codeintel/ReferencesPanel.tsx]: Yes, I should have done a better job of cleaning up the diff before requesting review 🤦🏻

File: client/web/src/codeintel/ReferencesPanel.tsx [hunk 5]
```diff
@@ -284,47 +289,62 @@ export const ReferencesList: React.FunctionComponent<
...
     // definitions) we select the first definition. We set it as activeLocation
     // and push it to the blobMemoryHistory so the code blob is open.
     useEffect(() => {
         if (props.jumpToFirst && definitions.length > 0) {
-            blobMemoryHistory.push(definitions[0].url, { syncToPanel: false })
-            setActiveLocation(definitions[0])
+            setActiveURL(definitions[0].url)
         }
-    }, [blobMemoryHistory, props.jumpToFirst, definitions])
+    }, [panelHistory, props.jumpToFirst, definitions, setActiveURL])
```
[the remark this answers is not in the document]
Comment (olafurpg) [reply client/web/src/codeintel/ReferencesPanel.tsx]: Good point, `panelHistory` has been removed. We still create an in-memory history inside `SideBlob` but it's no longer an abstraction we need to think about (or pass between components).

**Comment by olafurpg** [comment]
I've gone through the diff more carefully now and think it's ready to merge.

File: client/web/src/repo/blob/codemirror/linenumbers.ts [hunk 6]
```diff
@@ -159,12 +159,9 @@ export function selectableLineNumbers(config: {
             domEventHandlers: {
                 mousedown(view, block, event) {
                     const line = view.state.doc.lineAt(block.from).number
-                    const range = view.state.field(selectedLines)
 
                     view.dispatch({
-                        effects: (event as MouseEvent).shiftKey
-                            ? setEndLine.of(line)
-                            : setSelectedLines.of(isSingleLine(range) && range?.line === line ? null : { line }),
+                        effects: (event as MouseEvent).shiftKey ? setEndLine.of(line) : setSelectedLines.of({ line }),
```
Comment (olafurpg) [inline client/web/src/repo/blob/codemirror/linenumbers.ts:164]: FYI @fkling this PR includes minor fixes in the CodeMirror blob for issues I hit on. For example, I'm not sure what was the original intention with dispatching `setSeledctedLines.of(null)` and removing this fixed an interaction with the preview pane.

Comment (fkling) [reply client/web/src/repo/blob/codemirror/linenumbers.ts:164]: `setSeledctedLines.of(null)` causes the line to be deselected when it's clicked again.

**Comment by mrnugget** [comment]
Tested again, works fine locally! But the last-clicked-reference being selected on load doesn't seem to work consistently. Look at this video: the first few times I use the back button in the browser it seems to work, but then it doesn't.

https://user-images.githubusercontent.com/1185253/191249701-6b056dc6-1052-45d5-9e76-30aa5ea3cf9e.mp4

I'll try to find exact steps to reproduce, but so far no luck

**Comment by mrnugget** [comment]
Here's a shorter video, with debug view of session storage:

https://user-images.githubusercontent.com/1185253/191250738-1210a1ab-7135-4cba-88a5-316e273b41a3.mp4

File: client/web/src/repo/blob/codemirror/linenumbers.ts (hunk 6 above)
Comment (olafurpg) [reply client/web/src/repo/blob/codemirror/linenumbers.ts:164]: I reverted this change and added a special case for when `navigateToLineOnAnyClick` is true. Range selection and unselecting a line don't make sense when clicking on any line should navigate to that line.

**Comment by olafurpg** [comment]
@mrnugget That behavior is working as intended because you're promoting lines that don't appear in the collapsible list. We could try to preserve the scroll position somehow by finding the closest reference in the collapsible list, but I'd recommend we delay that for a separate PR because I don't anticipate this will be a big UX problem.

**Review by mrnugget — APPROVED** [review]

**Pull request merged**